##**Importing python libraries**

In [5]:
#Importing pandas, numpy, and datetime
import pandas as pd
import numpy as np
from datetime import datetime

##**Reading the CSV file**

In [26]:
##Reading csv file
dff = pd.read_csv(r'/content/Autolib_dataset (2).csv')
dff

Address  Cars  ...  hour  minute
0                    2 Avenue de Suffren     0  ...    11      43
1              145 Rue Raymond Losserand     6  ...     7      24
2       2 Avenue John Fitzgerald Kennedy     3  ...    20      14
3                 51 Rue Eugène Oudiné     3  ...     4      37
4     6 avenue de la Porte de Champerret     3  ...    17      23
...                                  ...   ...  ...   ...     ...
4995                 8 avenue Ménelotte     2  ...    11      26
4996                   37 rue de Dantzig     4  ...    16      56
4997                      142 rue du Bac     1  ...     7       1
4998         2 avenue du Val de Fontenay     2  ...    17      27
4999                 5 Rue Henry Brisson     4  ...    10       2

[5000 rows x 25 columns]

In [27]:
#displaying the first 10 rows
dff.head(10)

Address  Cars  Bluecar counter  ...  day  hour  minute
0                 2 Avenue de Suffren     0                0  ...    8    11      43
1           145 Rue Raymond Losserand     6                6  ...    6     7      24
2    2 Avenue John Fitzgerald Kennedy     3                3  ...    3    20      14
3              51 Rue Eugène Oudiné     3                3  ...    4     4      37
4  6 avenue de la Porte de Champerret     3                3  ...    8    17      23
5                8 Boulevard Voltaire     0                0  ...    6     7       2
6                      37 rue Leblanc     0                0  ...    8    18      20
7         17 Rue des Luaps Prolongée     3                3  ...    4    22      13
8               34 avenue Jean Moulin     1                1  ...    2    22      58
9        41 boulevard de Rochechouart     6                6  ...    4    15       2

[10 rows x 25 columns]

##**Getting information on the Data Set**

In [28]:
#getting dataset info
dff.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5000 entries, 0 to 4999
Data columns (total 25 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   Address              5000 non-null   object
 1   Cars                 5000 non-null   int64 
 2   Bluecar counter      5000 non-null   int64 
 3   Utilib counter       5000 non-null   int64 
 4   Utilib 1.4 counter   5000 non-null   int64 
 5   Charge Slots         5000 non-null   int64 
 6   Charging Status      5000 non-null   object
 7   City                 5000 non-null   object
 8   Displayed comment    111 non-null    object
 9   ID                   5000 non-null   object
 10  Kind                 5000 non-null   object
 11  Geo point            5000 non-null   object
 12  Postal code          5000 non-null   int64 
 13  Public name          5000 non-null   object
 14  Rental status        5000 non-null   object
 15  Scheduled at         47 non-null     object
 16  Slots 

##**Data Cleaning**

1.   Validity

                               1.1       Irrelevant columns

In [29]:
#before working on our data, we have to clean it
#the first step will be to remove the irrelevant columns

df_new = dff.drop(["Address", "Charge Slots", "Charging Status", "Displayed comment", "ID", "Geo point", "Rental status", "Slots", "Scheduled at", "Subscription status"], axis = 1, inplace = True) 


In [30]:
#we will then display the resulting data set
dff

Cars  Bluecar counter  Utilib counter  ...  day hour minute
0        0                0               0  ...    8   11     43
1        6                6               0  ...    6    7     24
2        3                3               0  ...    3   20     14
3        3                3               1  ...    4    4     37
4        3                3               0  ...    8   17     23
...    ...              ...             ...  ...  ...  ...    ...
4995     2                2               0  ...    6   11     26
4996     4                4               0  ...    4   16     56
4997     1                1               0  ...    1    7      1
4998     2                2               0  ...    4   17     27
4999     4                4               0  ...    8   10      2

[5000 rows x 15 columns]

2.       Outliers

In [31]:
#Checking for outliers using IQR scores
Q1 = dff.quantile(0.25)
Q3 = dff.quantile(0.75)

IQR = Q3 - Q1


(dff < (Q1 - 1.5 * IQR)) |(dff > (Q3 + 1.5 * IQR))

Bluecar counter   Cars   City   Kind  ...   hour  minute  month   year
0               False  False  False  False  ...  False   False  False  False
1               False  False  False  False  ...  False   False  False  False
2               False  False  False  False  ...  False   False  False  False
3               False  False  False  False  ...  False   False  False  False
4               False  False  False  False  ...  False   False  False  False
...               ...    ...    ...    ...  ...    ...     ...    ...    ...
4995            False  False  False  False  ...  False   False  False  False
4996            False  False  False  False  ...  False   False  False  False
4997            False  False  False  False  ...  False   False  False  False
4998            False  False  False  False  ...  False   False  False  False
4999            False  False  False  False  ...  False   False  False  False

[5000 rows x 15 columns]

3.       Accuracy

In [ ]:
# N/A

4.       Completeness

In [32]:
#Checking how many missing values there are across each column

dff.isnull().any()

Cars                  False
Bluecar counter       False
Utilib counter        False
Utilib 1.4 counter    False
City                  False
Kind                  False
Postal code           False
Public name           False
Station type          False
Status                False
year                  False
month                 False
day                   False
hour                  False
minute                False
dtype: bool

In [33]:
# Finding the total count of missing values

dff.isnull().sum().sum()

0

5.       Consistency

In [34]:
#Checking for and removing duplicates
#Checking for duplicates

dff[dff.duplicated()]

Empty DataFrame
Columns: [Cars, Bluecar counter, Utilib counter, Utilib 1.4 counter, City, Kind, Postal code, Public name, Station type, Status, year, month, day, hour, minute]
Index: []

In [35]:
#dropping duplicates

df2 = dff.drop_duplicates()

In [36]:
#checking for non duplicates
df2.head()

Cars  Bluecar counter  Utilib counter  ...  day hour minute
0     0                0               0  ...    8   11     43
1     6                6               0  ...    6    7     24
2     3                3               0  ...    3   20     14
3     3                3               1  ...    4    4     37
4     3                3               0  ...    8   17     23

[5 rows x 15 columns]

6.       Uniformity

In [ ]:
#N/A

In [38]:
# Creating a new column where we find the difference in the number of bluecars at the station/counter

df2['Bluecar_Diff'] = df2['Bluecar counter'].diff()
df2.head()

Cars  Bluecar counter  Utilib counter  ...  hour minute Bluecar_Diff
0     0                0               0  ...    11     43          NaN
1     6                6               0  ...     7     24          6.0
2     3                3               0  ...    20     14         -3.0
3     3                3               1  ...     4     37          0.0
4     3                3               0  ...    17     23          0.0

[5 rows x 16 columns]

**RESEARCH QUESTIONS**

# Identify the most popular hour of the day for picking up a shared electric car (Bluecar) in the city of Paris over the month of April 2018

In [57]:
# select the city Paris

df_paris = df2[df2['City'] =='Paris'] 

df_paris.head()

Cars  Bluecar counter  ...  Utilib_Diff  Utilib_Diff2
0     0                0  ...          NaN           NaN
1     6                6  ...          0.0           0.0
3     3                3  ...          1.0          -2.0
4     3                3  ...         -1.0           0.0
5     0                0  ...          0.0           0.0

[5 rows x 18 columns]

In [40]:
#counting and grouping the hours in descending order

df_paris[df_paris['Bluecar_Diff'] < 0].groupby('hour')['hour'].count().sort_values(ascending= False)

hour
21    80
16    68
7     64
13    62
5     62
23    61
20    61
3     60
8     59
4     58
18    57
9     57
17    55
19    55
22    55
14    54
12    54
0     52
10    52
2     51
6     50
15    49
11    47
1     43
Name: hour, dtype: int64

The most popular hour for pick up was 21

# What is the most popular hour for returning cars?

In [41]:
#counting and grouping the hours in descending order

df_paris[df_paris['Bluecar_Diff'] > 0].groupby('hour')['hour'].count().sort_values(ascending= False)

hour
3     49
12    48
10    48
6     46
8     44
21    43
11    42
19    42
15    42
18    40
9     40
0     39
22    38
2     38
1     38
16    37
5     37
23    35
13    33
7     33
17    33
4     31
20    27
14    25
Name: hour, dtype: int64

The most popular hour for return was 3

# What station is the most popular overall?

In [42]:
#We take the columns 'Kind' and 'Status' then group them by 'Public name', count them and group them in descending order
df_paris[(df_paris['Kind'] == 'STATION') & (df_paris['Status'] == 'ok')].groupby('Public name')['Public name'].count().sort_values(ascending= False)

Public name
Paris/Porte de Montrouge/8         13
Paris/Voltaire/182                 11
Paris/Courcelles/69                11
Paris/Mathis/35                    11
Paris/Philippe Auguste/126         11
                                   ..
Paris/Péguy/2                      1
Paris/Faubourg-Saint-Martin/168     1
Paris/Bobillot/16                   1
Paris/Fédération/10               1
Paris/Murat/157                     1
Name: Public name, Length: 518, dtype: int64

The most popular station overall is Porte de Montrouge

# What station is the most popular, at the most popular picking hour?

In [43]:
#We take the columns 'Kind', 'Status' and 'Hour', then group them by 'Public name' against the most popular picking hour, count them and group them in descending order
df_paris[(df_paris['Kind'] == 'STATION') & (df_paris['Status'] == 'ok') & (df_paris['hour'] == 21)].groupby('Public name')['Public name'].count().sort_values(ascending= False)

Public name
Paris/Tronchet/19               4
Paris/Voltaire/182              3
Paris/Émile Reynaud/4          2
Paris/Danton/8                  2
Paris/Boulard/1                 2
                               ..
Paris/Pelleport/86              1
Paris/Pereire/170               1
Paris/Pompe/137                 1
Paris/Pont Louis-Philippe/24    1
Paris/25 Août 1944/8           1
Name: Public name, Length: 106, dtype: int64

The most popular station at the most popular hour is Tronchet

#  What postal code is the most popular for picking up Bluecars overall?

In [44]:
df_paris[(df_paris['Kind'] == 'STATION') & (df_paris['Status'] == 'ok')].groupby('Postal code')['Postal code'].count().sort_values(ascending= False)

Postal code
75015    220
75016    197
75017    180
75011    155
75013    155
75008    150
75020    149
75012    138
75019    134
75014    128
75018    120
75007     99
75010     98
75009     93
75005     85
75006     74
75004     58
75003     55
75002     54
75001     43
75112      6
75116      5
Name: Postal code, dtype: int64

The most popular postal code overall is 75015

##Does the most popular station belong to that postal code?


No it does not

#  What postal code is the most popular for picking up Bluecars at the most popular hour?

In [45]:
df_paris[(df_paris['Kind'] == 'STATION') & (df_paris['Status'] == 'ok') & (df_paris['hour'] == 21)].groupby('Postal code')['Postal code'].count().sort_values(ascending= False)

Postal code
75014    15
75019    12
75012    11
75008    11
75016    10
75013     9
75011     8
75015     7
75020     5
75017     5
75007     5
75009     4
75006     3
75003     3
75010     3
75018     3
75005     2
75004     2
75001     2
75002     1
Name: Postal code, dtype: int64

The most popular postal code at the most popular hour is 75014

##**Using Utilib and Utilib 1.4 instead of Bluecar**

###**Utilib**

In [46]:
# Creating a new column where we find the difference in the number of Utilib at the station/counter

df2['Utilib_Diff'] = df2['Utilib counter'].diff()
df2.head()

Cars  Bluecar counter  Utilib counter  ...  minute Bluecar_Diff Utilib_Diff
0     0                0               0  ...      43          NaN         NaN
1     6                6               0  ...      24          6.0         0.0
2     3                3               0  ...      14         -3.0         0.0
3     3                3               1  ...      37          0.0         1.0
4     3                3               0  ...      23          0.0        -1.0

[5 rows x 17 columns]

In [49]:
# Identify the most popular hour of the day for picking up a shared electric car (Utilib) in the city of Paris over the month of April 2018

df_paris[df_paris['Utilib_Diff'] < 0].groupby('hour')['hour'].count().sort_values(ascending= False)

hour
19    12
23    10
22    10
14    10
6      9
10     9
5      8
8      8
9      8
3      7
11     7
15     6
16     6
17     6
7      6
20     6
2      6
13     5
4      5
0      5
12     4
1      4
18     3
21     3
Name: hour, dtype: int64

The most popular hour for Utilib pick up was 19

In [51]:
# Most popular hour for returning cars

df_paris[df_paris['Utilib_Diff'] > 0].groupby('hour')['hour'].count().sort_values(ascending= False)

hour
6     12
11     9
17     9
8      9
14     8
10     7
21     7
18     7
7      7
15     6
2      5
3      5
5      5
0      5
12     5
13     5
9      4
16     4
22     3
4      3
19     3
20     3
1      3
23     1
Name: hour, dtype: int64

The most popular hour for Utilib drop off was 06

In [68]:
# What station is the most popular? At the most popular picking hour?
  
df_paris[(df_paris['Kind'] == 'STATION') & (df_paris['Status'] == 'ok') & (df_paris['hour'] == 19)].groupby('Public name')['Public name'].count().sort_values(ascending= False)

Public name
Paris/Saint-Antoine/238    2
Paris/Nationale/77         2
Paris/Beaumarchais/87      2
Paris/Pyrénées/236       2
Paris/Clichy/181           2
                          ..
Paris/Pyrénées/89        1
Paris/Péguy/2             1
Paris/Richard Lenoir/63    1
Paris/Rungis/2             1
Paris/Adolphe Yvon/6       1
Name: Public name, Length: 89, dtype: int64

The most popular station at the most popular hour was Saint-Antoine

In [69]:
#  What postal code is the most popular for picking up Utilib Counter? Does the most popular station belong to that postal code?
# At the most popular picking hour?

df_paris[(df_paris['Kind'] == 'STATION') & (df_paris['Status'] == 'ok') & (df_paris['hour'] == 19)].groupby('Postal code')['Postal code'].count().sort_values(ascending= False)

Postal code
75020    9
75012    9
75016    8
75013    8
75011    8
75017    8
75015    6
75010    6
75003    6
75007    5
75014    4
75018    4
75019    4
75002    3
75008    2
75006    2
75004    2
75005    1
75001    1
Name: Postal code, dtype: int64

The most popular postal code for Utilib is 75020 at the most popular picking hour

###**Utilib 1.4**

In [54]:
# Creating a new column where we find the difference in the number of Utilib 1.4 counter at the station/counter

df2['Utilib_Diff2'] = df2['Utilib 1.4 counter'].diff()
df2.head()

Cars  Bluecar counter  ...  Utilib_Diff  Utilib_Diff2
0     0                0  ...          NaN           NaN
1     6                6  ...          0.0           0.0
2     3                3  ...          0.0           2.0
3     3                3  ...          1.0          -2.0
4     3                3  ...         -1.0           0.0

[5 rows x 18 columns]

In [58]:
# Identify the most popular hour of the day for picking up a shared electric car (Utilib 1.4 Counter) in the city of Paris over the month of April 2018

df_paris[df_paris['Utilib_Diff2'] < 0].groupby('hour')['hour'].count().sort_values(ascending= False)

hour
3     20
4     17
0     16
19    16
8     16
10    16
12    15
14    12
2     12
23    12
16    12
17    12
20    12
21    12
11    12
22    11
7     11
18    11
9     10
5     10
13     8
6      8
15     7
1      7
Name: hour, dtype: int64

The most popular hour for picking up Utilib 1.4 was 03

In [59]:
# What is the most popular hour for returning cars?

df_paris[df_paris['Utilib_Diff2'] > 0].groupby('hour')['hour'].count().sort_values(ascending= False)

hour
9     17
17    15
5     15
21    14
22    14
10    13
6     12
15    12
3     12
19    11
16    11
1     11
12    11
20    10
18    10
23     9
11     9
7      9
13     9
8      8
4      8
14     8
0      7
2      6
Name: hour, dtype: int64

The most popular hour for drop off was 09

In [65]:
# What station is the most popular? At the most popular picking hour?
  
df_paris[(df_paris['Kind'] == 'STATION') & (df_paris['Status'] == 'ok') & (df_paris['hour'] == 3)].groupby('Public name')['Public name'].count().sort_values(ascending= False)

Public name
Paris/Mortier/15             2
Paris/Julien Lacroix/39      2
Paris/Suffren/161            2
Paris/Quatre septembre/10    2
Paris/Francois 1er/7         2
                            ..
Paris/Pereire/115            1
Paris/Perrée/18             1
Paris/Plantes/56             1
Paris/Poissonnière/64       1
Paris/25 Août 1944/8        1
Name: Public name, Length: 89, dtype: int64

The most popular Station at the most popular hour was Mortier

In [71]:
#  What postal code is the most popular for picking up Utilib 1.4 Counter? Does the most popular station belong to that postal code?

# At the most popular picking hour?

df_paris[(df_paris['Kind'] == 'STATION') & (df_paris['Status'] == 'ok') & (df_paris['hour'] == 3)].groupby('Postal code')['Postal code'].count().sort_values(ascending= False)

Postal code
75016    9
75015    8
75020    8
75012    7
75018    7
75017    7
75011    6
75013    5
75008    5
75006    5
75014    4
75009    4
75007    4
75003    4
75019    3
75004    3
75002    2
75010    2
75116    1
75001    1
Name: Postal code, dtype: int64

The most popular postal code at most popular hour is 75016